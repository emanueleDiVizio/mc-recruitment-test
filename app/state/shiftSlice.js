import { createSlice } from '@reduxjs/toolkit'
import moment from 'moment'
const shiftSlice = createSlice({
  name: 'shifts',
  initialState: [],
  reducers: {
    startShift: state => [
      ...state,
      {
        startTime: moment().valueOf(),
        endTime: moment()
          .add(8, 'h')
          .valueOf(),
      },
    ],
    endShift: state => {
      var newState = [...state]
      return [{ ...newState.pop(), endTime: moment().valueOf() }, ...newState]
    },
  },
})

export default shiftSlice
