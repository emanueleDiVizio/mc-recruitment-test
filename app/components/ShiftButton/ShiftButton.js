import React from 'react'
import { useDispatch } from 'react-redux'
import { TouchableOpacity, Text, StyleSheet } from 'react-native'
import shiftSlice from '../../state/shiftSlice'
import useIsOnShift from '../../hooks/useIsOnShift'
import config from '../../config'

const ShiftController = () => {
  const dispatch = useDispatch()

  const [isOnShift] = useIsOnShift()

  const onUserPressButton = () => {
    isOnShift
      ? dispatch(shiftSlice.actions.endShift())
      : dispatch(shiftSlice.actions.startShift())
  }

  const computeStyles = onShift => ({
    backgroundColor: onShift ? config.colors.accent : config.colors.main,
  })

  return (
    <TouchableOpacity
      testID="shift-button"
      style={[computeStyles(isOnShift), styles.container]}
      onPress={onUserPressButton}
    >
      <Text style={styles.text}>{isOnShift ? 'End Shift' : 'Start shift'}</Text>
    </TouchableOpacity>
  )
}

const styles = StyleSheet.create({
  container: {
    height: 100,
    position: 'absolute',
    bottom: 0,
    right: 0,
    left: 0,
    shadowColor: 'black',
    shadowOpacity: 0.4,
    shadowRadius: 16,
    shadowOffset: { height: -4 },
  },
  text: {
    color: 'white',
    alignSelf: 'center',
    marginTop: 8,
    fontSize: 17,
  },
})

export default ShiftController
