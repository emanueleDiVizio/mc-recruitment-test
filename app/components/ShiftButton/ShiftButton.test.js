import React from 'react'
import ShiftButton from './ShiftButton'
import { fireEvent, render } from '@testing-library/react-native'
import * as redux from 'react-redux'
import shiftSlice from '../../state/shiftSlice'
import moment from 'moment'
import MockDate from 'mockdate'

describe('HandWashComponent', () => {
  let state
  let dispatch

  beforeEach(() => {
    dispatch = jest.fn()
    redux.useSelector = selector => selector(state)
    redux.useDispatch = () => dispatch
    MockDate.set('2018-1-1')
  })

  afterEach(() => {
    MockDate.reset()
  })

  it('renders correctly when off shift', () => {
    state = {
      shifts: [],
    }
    const { baseElement } = render(<ShiftButton />)

    expect(baseElement).toMatchSnapshot()
  })

  it('renders correctly when on shift', () => {
    state = {
      shifts: [{ endTime: moment().add(2, 'h') }],
    }
    const { baseElement } = render(<ShiftButton />)

    expect(baseElement).toMatchSnapshot()
  })

  it('dispatches `startShift` action when user presses button and user is not on shift', async () => {
    state = {
      shifts: [],
    }
    const { baseElement, getByTestId } = render(<ShiftButton />)

    const button = getByTestId('shift-button')
    fireEvent.press(button)

    expect(dispatch).toHaveBeenCalledWith(shiftSlice.actions.startShift())

    expect(baseElement).toMatchSnapshot()
  })

  it('dispatches `endShift` action when user presses button and is on shift', async () => {
    state = {
      shifts: [{ endTime: moment().add(2, 'h') }],
    }
    const { baseElement, getByTestId } = render(<ShiftButton />)

    const button = getByTestId('shift-button')
    fireEvent.press(button)

    expect(dispatch).toHaveBeenCalledWith(shiftSlice.actions.endShift())

    expect(baseElement).toMatchSnapshot()
  })
})
