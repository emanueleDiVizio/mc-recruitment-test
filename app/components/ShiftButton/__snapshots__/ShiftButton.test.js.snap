// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HandWashComponent renders correctly when off shift 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <
      style={
        Array [
          Object {
            "backgroundColor": "#00A778",
          },
          Object {
            "bottom": 0,
            "height": 100,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": -4,
            },
            "shadowOpacity": 0.4,
            "shadowRadius": 16,
          },
        ]
      }
      testID="shift-button"
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "white",
            "fontSize": 17,
            "marginTop": 8,
          }
        }
      >
        Start shift
      </Text>
    </>
  </View>
</View>
`;

exports[`HandWashComponent renders correctly when on shift 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <
      style={
        Array [
          Object {
            "backgroundColor": "#00A1D2",
          },
          Object {
            "bottom": 0,
            "height": 100,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "shadowColor": "black",
            "shadowOffset": Object {
              "height": -4,
            },
            "shadowOpacity": 0.4,
            "shadowRadius": 16,
          },
        ]
      }
      testID="shift-button"
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "white",
            "fontSize": 17,
            "marginTop": 8,
          }
        }
      >
        End Shift
      </Text>
    </>
  </View>
</View>
`;
